### ISSUE TITLE
The usage of C++ <random> header

### ISSUE THREAD

**Issue by LesleyLai** [RayTracing/raytracing.github.io#468]
In the "in One Weekend" book chapter 7, the book introduced the usage of the `<random>` header. The current function looks like this:

```cpp
inline double random_double_2() {
  static std::uniform_real_distribution<double> distribution(0.0, 1.0);
  static std::mt19937 generator;
  static std::function<double()> rand_generator = 
    std::bind(distribution, generator);
  return rand_generator();
}
```

The usage of `std::function` and `std::bind` are unnecessary, and the book also does not have any explanation on it. The following function should be functional equivalent, but also easier to understand for people not familiar with those standard library constructs:

```cpp
inline double random_double_2() {
  static std::uniform_real_distribution<double> distribution(0.0, 1.0);
  static std::mt19937 generator;
  return distribution(generator);
}
```

If you want me to change that, I can create a PR.

**Comment by trevordblack** [comment]
This part of chapter 7 is supposed to be an aside that tells the reader where to find an RNG if `rand()` isn't working (for _some_ reason). I don't think we ever refined it.

I've never worked in `<random>`. Does the `return distribution(generator)` remain static and has the same seed across runs?

**Comment by LesleyLai** [comment]
The seed is given at `static std::mt19937 generator;`, so even the current code in the book does not have the same seed across runs. If you want the same seed, you need to do `static std::mt19937 generator(seed);`.

I don't understand the meaning of your last sentence. I am not proposing to remove the `rand` part.

**Comment by hollasch** [comment]
Hey Lesley, go ahead and submit a PR to the `dev-patch` branch. Can you add some explanation for both seeded and unseeded runs? Also, please add an entry to the CHANGELOG. Thanks!

**Comment by trevordblack** [comment]
Following up on this.

Are you able to find time to close this out?

**Comment by LesleyLai** [comment]
Sorry, I am busy with my final exam for these two weeks. I will have a PR next week after it is over.

**Comment by hollasch** [comment]
No worries, it's been moved to v3.1.1 or v3.2.0.

**Comment by hollasch** [comment]
I think we need to provide a way to start with a specific seed. Lesley, your comment above suggests that the current PRNG _cannot_ be seeded — is that correct?

**Comment by LesleyLai** [comment]
The current book doesn't do any seeding, but it can be easily done. The C standard `rand()` can be seeded by another function `srand()`. For the C++ `<random>` header, a seed can be provided when initializing the generator, otherwise, the seed is provided by the system.

Also, I will finish my final tomorrow and then I will have some time to write a PR :smile:.

**Comment by trevordblack** [comment]
`rand()` has the default seed set to 1:

https://www.cplusplus.com/reference/cstdlib/srand/
```
If seed is set to 1, the generator is reinitialized to its initial value and produces the same values as before any call to rand or srand.
```

A call to `rand()` without a call to `srand(int)` will always produce the same stream of integers for a specific build environment. I'm not sure how the produced values of `rand` changes across different compilers and operating systems.

If we want the reader to be able to match the book output at a per-pixel basis, we _might_ need to explicitly call to `srand` with a seed, e.g. `srand(0x1bad5eed)`. But if all we want is for the user to have consistent streams of random numbers, then the code as-is (i.e. no call to `srand`) is doing that.

I have to relearn all this stuff for the Multithreading chapter

**Comment by LesleyLai** [comment]
`rand()` use implementation-defined algorithms, so there is no way to guarantee same stream of integers across different compilers. It is also not thread-safe, so better avoided in multi-threaded rendering.

**Comment by peteroupc** [comment]
Although `rand` behaves by default as though `srand(1)` were called, the C standard doesn't specify what algorithm `rand()` has to use, leaving that detail up to the C implementation.  Thus, the results will not be consistent between C library implementations.

**Comment by peteroupc** [comment]
> `rand()` use implementation-defined algorithms, so there is no way to guarantee same stream of integers across different compilers.

Note that `uniform_real_distribution` (and other distribution classes) have the same problem, unlike with random engines like `mt19937`.

>  It is also not thread-safe, so better avoided in multi-threaded rendering.

 C++ distributions have this problem too, as do the standard C++ engines.

**Comment by LesleyLai** [comment]
> 
> 
> > `rand()` use implementation-defined algorithms, so there is no way to guarantee same stream of integers across different compilers.
> 
> Note that `uniform_real_distribution` (and other distribution classes) have the same problem, unlike with random engines like `mt19937`.

Exactly, sometimes you must use your own RNG facilities.

> 
> > It is also not thread-safe, so better avoided in multi-threaded rendering.
> 
> C++ distributions have this problem too, as do the standard C++ engines.

You can declare C++ distributions and C++ engines as `thread_local` or local variables.

**Comment by hollasch** [comment]
Addressed in PR #514
